### PR TITLE
`PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` migration

### LINKED ISSUE #13343: Migrate running process instance into `PROCESS_INSTANCE_KEYS_BY_DEFINITION_KEY` ColumnFamily
## Description
This ColumnFamily must contains **all** running process instances. We must create a migration script that initially fills this ColumnFamily.

- Create the migration script
    - Include tests!

**Blocked by**
#13340

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This migration does the initial fill of the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` ColumnFamily. It iterates over
all element instances in the state. It checks if it is of type `PROCESS` and if it is it will insert it with the corresponding process definition key in the new ColumnFamily.

## Related issues

closes #13343 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
Can be merged after #13342

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @remcowesterhoud

❌ We need to evaluate whether this migration will take too much time on clusters with large state.

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java
Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:339]: > would be willing to share his thoughts on how large our states can be nowadays.

That totally depends on the workload, so not sure how to answer this :)

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:339]: @Zelldon Perhaps you can share the maximum you encountered in your recent tests. It may be a ballpark number for us to use in testing how long migration of such a large state would take.

Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:339]: Still, this question is not clear to me. It can be as large as we want to go 😅 

In one of my tests for example I had a 80 gig snapshot. Filled with 3 partitions a 500 gig disk. 

In previous versions, we had issues with 1 gig of snapshot, due to the performance bottleneck for RocksDB iterations, which should now be solved. 

I can't answer what a good size is since this totally depends on the use case and the customer. Might make sense to check in SaaS what the avg snapshot size is. Or you just go with 5 to 10 gig 🤷🏼 but then this is just a guess again. I would say at least go over the 1 gig.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java:339]: Thanks for this discussion. I will leave the PR open for now and run some tests once I've progressed further along with the deletion topic . Like @korthout I'm also not sure how I could improve this further currently.

**Comment by remcowesterhoud** [comment]
I ran a benchmark with a migrations that just did a sleep of 10 minutes:

image

The screenshot shows that the cluster was started at ~15:53. As expected it took about 10 minutes for processing to start at ~16:05.
The cluster achieved its full potential after ~20 minutes. One of the partitions changed leader and this leader had to redo the migration, adding another 10 minutes before this partition started processing.

**Comment by remcowesterhoud** [comment]
After a [discussion on Slack](https://camunda.slack.com/archives/C03JKGAE8D7/p1690901523632939) we decided it is acceptable to merge this, as we don't promise any SLA on updating Zeebe.

A better solution for this will be found as part of https://github.com/camunda/product-hub/issues/256.
I have created an issue to document the impact this can have in the upgrade guide: https://github.com/camunda/camunda-platform-docs/issues/2384

@korthout could I get your approval on this PR so I can merge it?

**Review by korthout — APPROVED** [review]
Thanks for benchmarking and following up on this risk @remcowesterhoud 

LGTM 👍

**Pull request merged**
